### PR TITLE
:seedling: TK-2662: Added More Resources, Updated Doc

### PR DESCRIPTION
**Pull request by pallav-trilio**
Added Resources related to Cluster such as Node, Namespaces and Few Namespaced resources such as ResourceQuota and LimitRange

### CODE REVIEW

File: tools/log-collector/log_collector.go [hunk 1]
```diff
@@ -217,7 +194,7 @@ func (l *LogCollector) getResourceObjects(resourcePath string, resource *apiv1.A
 			listPath := fmt.Sprintf("%s/namespaces/%s/%s", resourcePath, l.Namespaces[index], resource.Name)
 			err := l.disClient.RESTClient().Get().AbsPath(listPath).Do(context.TODO()).Into(&obj)
 			if err != nil {
-				if errors.IsNotFound(err) || errors.IsForbidden(err) {
+				if apierrors.IsNotFound(err) || apierrors.IsForbidden(err) {
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:197]: log the `err`. 
Review if it's missing from other places.. must log error for all API calls,

File: tools/log-collector/log_collector.go [hunk 2]
```diff
@@ -748,3 +734,66 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
+		return cgErr
+	}
+	namespaceResource := getResourceByName(coreGV, Namespaces)
+	namespaceObjects := l.getResourceObjects(getAPIGroupVersionResourcePath(CoreGv), &namespaceResource)
+	allNamespaces := getObjectsNames(namespaceObjects)
+
+	if len(l.Namespaces) != 0 && !l.isSubset(allNamespaces) {
+		err := errors.New("specified namespaces doesn't exists in the cluster")
+		return err
+	}
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:791]: this check should be set first.. then nodes and other resources should be collected.

File: tools/log-collector/log_collector.go [hunk 3]
```diff
@@ -33,49 +35,24 @@ type LogCollector struct {
...
-	allNamespaces := getObjectsNames(namespaceObjects)
+	l.setClient()
+	l.getClusterNodes()
 
-	if len(l.Namespaces) != 0 && !l.isSubset(allNamespaces) {
-		log.Error("Specified namespaces doesn't exists in the cluster")
+	nsErr := l.checkNamespaces()
+	if nsErr != nil {
+		log.Errorf("%s", nsErr.Error())
 		return nil
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:53]: why are you returning `nil` here?

Also, log error of `CollectLogsAndDump` func.

File: tools/log-collector/log_collector.go [hunk 4]
```diff
@@ -748,3 +734,66 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
 }
+
+func (l *LogCollector) getClusterNodes() {
+
+	var nodeList unstructured.UnstructuredList
+	gvk := schema.GroupVersionKind{Kind: NodeKind, Version: CoreGv}
+	nodeList.SetGroupVersionKind(gvk)
+	err := l.k8sClient.List(context.Background(), &nodeList)
+	if err != nil {
+		log.Error("Unable to Fetch Node List")
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:745]: return with err, from here

File: tools/log-collector/log_collector.go [hunk 5]
```diff
@@ -748,3 +734,66 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
+	}
+	for _, node := range nodeList.Items {
+		resourceDir := filepath.Join(node.GetKind())
+		yErr := l.writeYaml(resourceDir, node)
+		if yErr != nil {
+			log.Error("Unable to Write Node Yaml")
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:751]: return with err, from here

File: tools/log-collector/log_collector.go [hunk 6]
```diff
@@ -748,3 +734,66 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
+	log.Info("Checking Namespaces")
+	coreGV, cgErr := l.getAPIGVResources(CoreGv)
+	if cgErr != nil {
+		return cgErr
+	}
+	namespaceResource := getResourceByName(coreGV, Namespaces)
+	namespaceObjects := l.getResourceObjects(getAPIGroupVersionResourcePath(CoreGv), &namespaceResource)
+	allNamespaces := getObjectsNames(namespaceObjects)
+
+	if len(l.Namespaces) != 0 && !l.isSubset(allNamespaces) {
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:788]: optimise func `isSubset` ... use `sets` data type and it's native funcs

**Review by msurudwad — CHANGES_REQUESTED** [review]

**Review by msurudwad — CHANGES_REQUESTED** [review]
func `getResourceByName` can be removed, if you maintain map and return that map from func `getAPIGVResources` instead of arrray.

File: tools/log-collector/log_collector.go (hunk 1 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:197]: resolved

File: tools/log-collector/log_collector.go (hunk 2 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:791]: resolved

File: tools/log-collector/log_collector.go (hunk 3 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:53]: resolved

File: tools/log-collector/log_collector.go (hunk 4 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:745]: resolved

File: tools/log-collector/log_collector.go (hunk 5 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:751]: resolved

File: tools/log-collector/helper.go [hunk 7]
```diff
@@ -316,6 +321,8 @@ func filterGroupResources(resources []apiv1.APIResource, group string) (filtered
 			filteredResources = append(filteredResources, resources[index])
 		} else if group == BatchGv1beta1 {
 			filteredResources = append(filteredResources, resources[index])
+		} else if group == extension {
+			filteredResources = append(filteredResources, resources[index])
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply tools/log-collector/helper.go:325]: resolved

File: docs/log-collector/README.md [hunk 8]
```diff
@@ -78,5 +93,13 @@ StatefulSet
...
 PersistentVolumes  
 Services  
+Endpoints
+Ingress
+Events
+ConfigMap
+LimitRange
+ResourceQuota
+Namespaces
+Nodes
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply docs/log-collector/README.md:103]: resolved

File: tools/log-collector/log_collector.go (hunk 6 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:794]: resolved..made changes from slice to map and reduced the loop. in our cases there are multiple namespaces to be selected from a array of namespaces

File: tools/log-collector/log_collector.go (hunk 1 above)
Comment (msurudwad) [reply tools/log-collector/log_collector.go:204]: If `err` is not `apierrors.IsNotFound(err)` or `apierrors.IsForbidden(err)` then, you should log and return err

File: tools/log-collector/log_collector.go [hunk 9]
```diff
@@ -797,3 +810,24 @@ func (l *LogCollector) checkNamespaces() error {
 
 	return nil
 }
+
+// getApiGVResource returns resource for given resourceName
+func (l *LogCollector) getAPIGVResource(apiGroupVersion, resourceName string) (gVResources apiv1.APIResource, err error) {
+
+	var gVResourcesList *apiv1.APIResourceList
+	gVResourcesList, err = l.disClient.ServerResourcesForGroupVersion(apiGroupVersion)
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:818]: You'll have make `disClient` calls for same group multiples times if different `resourceName` is passed to func `getAPIGVResource`. Make one  `disClient` call for one group and add it map and use it later for different `resourceName` [no need to call func `getApiGVResource` everytime].

File: tools/log-collector/log_collector.go [hunk 10]
```diff
@@ -338,11 +347,7 @@ func (l *LogCollector) writeLogs(resourceDir string, obj unstructured.Unstructur
 }
 
 // isSubset checks whether the given namespaces is a subset of all Namespaces in cluster
-func (l *LogCollector) isSubset(second []string) bool {
-	set := make(map[string]string)
-	for _, value := range second {
-		set[value] = value
-	}
+func (l *LogCollector) isSubset(set map[string]string) bool {
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:350]: Good for now.. next time, try to use `sets` data type.. there are ready funcs like `IsSuperset` which can used directly.. no need to write custom logic.

File: tools/log-collector/helper.go [hunk 11]
```diff
@@ -34,20 +34,20 @@ const (
...
-	extension     = "extensions/v1beta1"
+	networkingGv  = "networking.k8s.io/v1beta1"
 
 	Namespaces          = "namespaces"
 	Events              = "events"
 	CRD                 = "customresourcedefinitions"
 	StorageClass        = "storageclasses"
 	VolumeSnapshot      = "volumesnapshots"
 	VolumeSnapshotClass = "volumesnapshotclasses"
-	ConversionNamespace = "trilio-conversion"
```
Comment (msurudwad) [inline tools/log-collector/helper.go:44]: is this not needed? are we not collecting resources from `ConversionNamespace`?

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: tools/log-collector/helper.go (hunk 11 above)
Comment (pallav-trilio) [reply tools/log-collector/helper.go:44]: Conversion Namespace is now deprecated

File: tools/log-collector/log_collector.go (hunk 10 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:350]: Alright

File: tools/log-collector/log_collector.go [hunk 12]
```diff
@@ -217,7 +201,8 @@ func (l *LogCollector) getResourceObjects(resourcePath string, resource *apiv1.A
 			listPath := fmt.Sprintf("%s/namespaces/%s/%s", resourcePath, l.Namespaces[index], resource.Name)
 			err := l.disClient.RESTClient().Get().AbsPath(listPath).Do(context.TODO()).Into(&obj)
 			if err != nil {
-				if errors.IsNotFound(err) || errors.IsForbidden(err) {
+				if apierrors.IsNotFound(err) || apierrors.IsForbidden(err) {
+					log.Errorf("%s", err.Error())
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:205]: log this on `warning` level.  Wherever `err` is ignored but needs to be logged then log it at Info or `warning` level.

File: docs/log-collector/README.md [hunk 13]
```diff
@@ -102,4 +102,10 @@ ResourceQuota
...
 Nodes
 ```  
-and ```TrilioVault Resources```
\ No newline at end of file
+and ```TrilioVault Resources```
+
+## OCP Specific Resources Considered for Log Collection:
+
+```  
+ClusterServiceVersion
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply docs/log-collector/README.md]: Will creata a new ticket for that

File: tools/log-collector/log_collector.go (hunk 9 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:818]: `getAPIGVResource` and `getAPIGVResources`  are two different for group and specific resource

File: tools/log-collector/log_collector.go (hunk 12 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:205]: resolved

File: tools/log-collector/helper.go [hunk 14]
```diff
@@ -242,9 +231,10 @@ func getContainerStatusValue(containerStatus *corev1.ContainerStatus) (conStatOb
 }
 
 // getObjectsNames returns list of names of objects
-func getObjectsNames(objects unstructured.UnstructuredList) (nameList []string) {
+func getObjectsNames(objects unstructured.UnstructuredList) map[string]string {
+	nameList := make(map[string]string)
 	for index := range objects.Items {
-		nameList = append(nameList, objects.Items[index].GetName())
+		nameList[objects.Items[index].GetName()] = objects.Items[index].GetName()
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply tools/log-collector/helper.go:237]: Just for checking the value present in that. More effective than looping

File: tools/log-collector/log_collector.go [hunk 15]
```diff
@@ -748,3 +714,132 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
 	}
 	return objects
 }
+
+func (l *LogCollector) getClusterNodes() error {
```
Comment (madhurtrilio) [inline tools/log-collector/log_collector.go:718]: Function name should be getNodeInfo

File: tools/log-collector/log_collector.go [hunk 16]
```diff
@@ -748,3 +714,132 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
+			log.Error("Unable to Write Node Yaml")
+			return yErr
+		}
+	}
+
+	return nil
+}
+
+// WriteNs writes yaml of requested namespaces
+func (l *LogCollector) WriteNs(namespaceObjects unstructured.UnstructuredList) error {
```
Comment (madhurtrilio) [inline tools/log-collector/log_collector.go:741]: getNamespaceInfo

File: tools/log-collector/log_collector.go [hunk 17]
```diff
@@ -748,3 +714,132 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
+			if eLrr != nil {
+				return eLrr
+			}
+		}
+	}
+	return nil
+}
+
+// checkNamespaces taken all given ns from user and checks the same in cluster and write it YAML's
+func (l *LogCollector) checkNamespaces() error {
```
Comment (madhurtrilio) [inline tools/log-collector/log_collector.go:762]: checkNamespaces should be a separate function to just check if the namespaces exist and should be called at the start. If the given namespace list is not correct then we should error out at the very beginning

File: tools/log-collector/log_collector.go [hunk 18]
```diff
@@ -748,3 +714,132 @@ func (l *LogCollector) getResourceObjectsWithOwnerRef(resourcePath string,
...
+					return gVResourcesList.APIResources[index], nil
+				}
+			}
+		}
+	}
+	return gVResource, nil
+}
+
+// checkingStorageGroup get all storage classes and Write it's YAML
+func (l *LogCollector) checkingStorageGroup() error {
```
Comment (madhurtrilio) [inline tools/log-collector/log_collector.go:807]: getStorageClassInfo

**Review by madhurtrilio — CHANGES_REQUESTED** [review]
The function names here are all messed up. Can't figure out from any function what that function is performing

**Pull request closed without merging**